Enhance API data import documentation with advanced content

Endalaner opened the pull request:
- Added a warning section for advanced users, clarifying the intended audience for the API data import guide.
- Included alternative data import methods for beginners, linking to GUI and Python script options.
- Expanded the data import workflow with detailed steps for importing studies, samples, and linking data using async job endpoints
- Introduced new GIFs to visually demonstrate API interactions and improve user understanding.
- Updated the content to reflect recent changes in API endpoints and data handling processes.

Related support ticket - https://genestack.atlassian.net/browse/SUP-1618

Commit:
Enhance API data import documentation with advanced content and new examples

- Added a warning section for advanced users, clarifying the intended audience for the API data import guide.
- Included alternative data import methods for beginners, linking to GUI and Python script options.
- Expanded the data import workflow with detailed steps for importing studies, samples, and linking data using async job endpoints
- Introduced new GIFs to visually demonstrate API interactions and improve user understanding.
- Updated the content to reflect recent changes in API endpoints and data handling processes.

Commit:
- Replaced outdated step-by-step examples with new sample and library-based import workflow links in the user guide.
- Updated the Metadata Editor documentation to include a missing GIF for better visual guidance on using the dashboard.

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -19,23 +29,30 @@ You can optionally also import:
 
-You should provide HTTPS file links (which can include authentication credentials) to a storage that’s accessible
-by ODM. File uploads from a local computer will be supported in a future release.
+API allows loading files hosted at FTP or HTTP web addresses or contained in a mounted ODM NFS storage.
```
I am under the impression that for the import/job APIs 3 sources can be used:
 - HTTP URLs,
 - S3 URIs,
 - NFS for mounted storage.
 
 Could you please check with the development team to see if the FTP is supported?

Endalaner replied:
I have checked with a public FPT file:
`{
  "error": {
    "class": "com.genestack.api.exceptions.GenestackClientException",
    "message": "Incorrect protocol in importParams.signalFileUrl: ftp Should be one of [https, http, s3, file]",`

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/doc-odm-user-guide/images/data-model+metainfo-editor.png:
Let's use an up-to-date screenshot!
![Uploading Screenshot_1.png…]()

Endalaner replied:
Done

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -115,6 +122,15 @@ tokens.
+!!! note "Token compatibility" 
+    The solution has been tested with the Azure AD access tokens only. For other providers pretesting is recommended.
+
 You could also be provided with an Access Token. To use it, in the follow examples replace the authorization header part
```
Should this old sentence be removed, as it was replaced with the more detailed description above?

Endalaner replied:
I did not get this one. Can you please explain in a little more details?

Endalaner replied:
Removed

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +143,370 @@ with
+
+![api-navigate-swagger.gif](doc-odm-user-guide/gifs/api-navigate-swagger.gif)
+
+For studies,
```
Now, all data loading endpoints are grouped - users don't have to switch between studyCurator, sampleCurator, etc. I would suggest expressing it in the updated description:

"For data import, you should go to the *job* section and choose the endpoint relevant for the specific data type. For studies, use the **POST /api/v1/jobs/import/study** method, and supply the file URL:"

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -127,210 +143,370 @@ with
+```
+
+!!! note "templateId"
+    By the default "templateId" is included in the call, it specifies  the template that will be assigned to your new study. If you are not sure what is the correct template accession - you can remove this field from the call, the default template will be assigned to your study. To specify the template that you would like to use - you will need to supply the correct template accession, that can b ...
````
I would rather avoid stating that "templateId" is included in the call by the default - it is included in the example in Swagger. 

We can state that: "You can include an optional parameter "templateId" to specify which template should be associated with the loaded data. You will need to provide the accession of the desired template, which can be obtained the Template Editor. If the "templateId" parameter is not specified, the default template set for the instance will be used."

Endalaner replied:
Agree, done

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +143,370 @@ with
 
-1. **Import samples**
+#### Working with the jobExecId
+The following endpoints allow you to manage and inspect jobs using the jobExecId, which is returned after initiating an import or other asynchronous task.
```
Let's use "after initiating an asynchronous import task."

We don't have any other asynchronous tasks there at the moment

Endalaner replied:
Agree, done

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +143,370 @@ with
-(with accession GSF282799):
+You can link samples to study using the integration endpoint **POST /api/v1/as-curator/integration/link/sample/group/{sourceId}/to/study/{targetId}**, specifying the accessions of the pair of objects to be
+linked. The following call will link samples that we imported in the previous step (with accession GSF1283530) to the study
+(with accession GSF1283528):
```
"The pair of objects" is not fully correct. When we use the group linking endpoint, we link all samples from the file to the study in one go, so we link number of objects to the study.
 
 "You can link samples to study using the integration endpoint **POST /api/v1/as-curator/integration/link/sample/group/{sourceId}/to/study/{targetId}**, specifying the accessions of the study and the accession of the sample group. This will link all samples from the imported file to the study. The following call will link samples that we imported in the previous step (with accession GSF1283530) to the study
(with accession GSF1283528):"

Endalaner replied:
Agree, done

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +143,370 @@ with
+![sample_added.gif](doc-odm-user-guide/gifs/sample_added.gif)
 
-You can link the other samples in the same way. In practice, we’d create a script to automate this process, using common sample IDs.
+You can link other samples in the same way. This example demonstrates a simple procedure for linking two entities, which can serve as a foundation for building automated import pipelines.  
```
I think this part in the old guidance was added to explain that you do not need to manually link every single sample from the file by the ID, but rather create a script. Now we do have this call for group linkage,we can get rid of it.

We may say that "Samples from other files can be loaded in the same way. They will be displayed in the Metadata Editor on a separate tab.

Endalaner replied:
Agree, done

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +143,370 @@ with
 
-You can then link these run-level expression objects to the corresponding samples. The call below will link
-the expression object GSF282812 to the sample HG00119 (that’s been linked to the study in the previous step):
+You can then link this expression either by group (link expression group to sample group) or object to object (expression object to specific sample). 
```
Let`s remove the part about linking single objects - this were added for Roche and not currently in use.

Endalaner replied:
Agree, done.

eeliane replied:
I still see information regarding linking object to object, are we finally going to remove?

Endalaner replied:
@MariaBorodaenko @eeliane @MikhailAf Let's decide in this thread about mentioning single object linking. I am fine with both options, but if we are not promoting this feature to the clients - it is a good idea to remove it from the user guide.

eeliane replied:
I agree that we need to decide since I didn't see this thread initially and asked @Endalaner to describe both approaches https://github.com/genestack/user-docs/pull/120#discussion_r2069094807

Endalaner replied:
Decided on the call with @eeliane and @MikhailAf that we will keep it in the user guide.

MariaBorodaenko commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +143,370 @@ with
+
+### Import and link variant data to samples
+
+Let’s repeat the previous step. This time, for variant data, so that the same sample group is linked to both expression and variant data.
```
I would suggest a slight rewording to highlight the hierarchy in our data model: expression and variant data are to be linked to samples. Not vice versa 

Samples are linked to a study.

Endalaner replied:
> Let's repeat the previous step, this time for variant data, ensuring that both expression and variant data are linked to the samples, reinforcing the data model hierarchy where samples are linked to a study, and data types (expression and variant) are linked to samples

For review.

Commit:
- Modified the file import sources list to include HTTP/HTTPS URLs, S3 URIs, and NFS paths and exclude FTP.
- Clarified instructions for linking samples to studies and managing job execution IDs.
- Improved the explanation of the optional "templateId" parameter for data imports.
- Updated various sections for better readability and consistency.

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -1,3 +1,13 @@
+!!! warning "Advanced Content – API Data Import"
```
I suggest to use another admonition instead of warning message:
<img width="916" alt="image" src="https://github.com/user-attachments/assets/daadbb6c-7458-4d39-a192-74b651d45288" />

Endalaner replied:
I tried "note" and "info" but the style did not match with the one that you show on the screenshot. Can you tell what type is that?
**NVM found it.**

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -1,3 +1,13 @@
+
 # Import Data Using API
 
 On this page, you will find a brief explanation about data import via APIs, followed by a step-by-step example that
```
We need to update info regarding ALL the entities to be imported via API, not only Study and Samples

eeliane replied:
I agree, since we started to update this part of our user guide, currently we have nothing about preparations, flow cytometry, attached files, gtm (maybe can be skipped, I'm not sure). I'd also mention that in libs/preps files should be provided Sample Source ID column with IDs that will be used to link libs/preps to samples. And that to link data to libs/preps only default attribute can be used.

Endalaner replied:
I have added info on preparations, flow cytometry, attached file. Info on libs\preps mandatory attribute, default link attribute for libs and preps

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -19,23 +29,31 @@ You can optionally also import:
 - *Cross-reference mapping*: a list of transcript and gene ids and how they map to each other.
 - *Libraries metadata*: metadata about sample library preparation for transcriptomics data.
```
I suggest not to place Libs and Preps as _optional_. Let's have just a single list of everything

MikhailAf replied:
For any data that can be indexed, let's avoid mentioning just transcriptomics

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -19,23 +29,31 @@ You can optionally also import:
 - *Cross-reference mapping*: a list of transcript and gene ids and how they map to each other.
 - *Libraries metadata*: metadata about sample library preparation for transcriptomics data.
 - *Preparations metadata*: metadata about sample preparation for proteomics data.
+- *Files*: Supplement your study by attaching related research materials like PDF, XLSX, DOCX, PPTX files, images, and more. Please note, contents of these attached files won't be indexed or made searchable.
```
Files -> Attached Files (we call them that way to prevent miscommunication)

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -19,23 +29,31 @@ You can optionally also import:
 - *Cross-reference mapping*: a list of transcript and gene ids and how they map to each other.
 - *Libraries metadata*: metadata about sample library preparation for transcriptomics data.
```
Definition for Library as "sample library preparation" sounds strange to me

Endalaner replied:
Can you share your idea on what would be a better description? I have checked the user guide for a better option, but was not able to find one.

eeliane replied:
@Endalaner  could you please check our presentations, jira tickets and confluence pages to offer better definitions?

Endalaner replied:
Updated

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -19,23 +29,31 @@ You can optionally also import:
 - *Cross-reference mapping*: a list of transcript and gene ids and how they map to each other.
 - *Libraries metadata*: metadata about sample library preparation for transcriptomics data.
 - *Preparations metadata*: metadata about sample preparation for proteomics data.
```
Any data, not only proteomics

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -19,23 +29,31 @@ You can optionally also import:
 
 Importing data has two stages. First, you import studies, samples, and data separately. Then, you link them
-together: a study can be linked to multiple samples and a sample can be linked to multi-omics or other types of data. The **Sample Source ID** is used as the default linking key. You can choose another attribute from the template for linking data to samples.. The data model and how it looks in the User Interface is shown below:
+together: a study can be linked to multiple samples and a sample can be linked to multi-omics or other types of data. The **Sample Source ID** is used as the default linking key. You can choose another attribute from the template for linking data to samples. The data model and how it looks in the User Interface is shown below:
```
I think it worth mentioning also Libs and Preps and that they cannot be linked by any attribute and require their own ID that is linked to SS ID

eeliane replied:
I believe we need to mention a proper and correct order of linking, based on our endpoints: we link samples TO study, libs/preps TO samples, omics data TO samples/libs or preps, and attached files TO study.

Endalaner replied:
My proposal:

> Importing data has two stages. First, you import studies, samples, and data separately. Then, you link them together: a study can be linked to multiple samples, and a sample can be linked to multi-omics or other types of data. The **Sample Source ID** is used as the default linking key. You can choose another attribute from the template for linking data to samples. The data model and how it looks in the User Interface is shown below.
> 
> In addition to core data types, **libraries** and **preparations** require special handling. Unlike omics data, they cannot be linked using arbitrary attributes — they must include their own unique ID that corresponds to the **Sample Source ID**.
> 
> The correct order of linking follows the system logic and available endpoints:
> 
> - **Samples** are linked to a **study**
> - **Libraries** and **preparations** are linked to **samples**
> - **Omics data** (e.g. transcriptomics, proteomics) are linked to **samples**, or to **libraries/preparations** depending on the data type
> - **Attached files** are linked directly to a **study**

eeliane replied:
Could you please fix this part and use proper order of linking: 
>a study can be linked to multiple samples, and a sample can be linked to multi-omics or other types of data.


This part about linking libs/preps to samples and required attributes isn't clear also, could you please mention required attributes in the files
>In addition to core data types, libraries and preparations require special handling. Unlike omics data, they cannot be linked using arbitrary attributes — they must include their own unique ID that corresponds to the Sample Source ID.

It depends not on data type but on user goal/user study structure, could you please mention what data can be linked to samples/libs/preps and what data can't be linked to libs and preps?
>Omics data (e.g. transcriptomics, proteomics) are linked to samples, or to libraries/preparations depending on the data type

Endalaner replied:
Agree, added.

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -19,23 +29,31 @@ You can optionally also import:
+API allows loading files hosted at HTTP/HTTPS URLs, S3 URIs, and NFS paths for files stored in mounted ODM storage.
 
-## Step-by-step example 1
+!!! note "Important"
```
Let's make it as Limitation here:
<img width="894" alt="image" src="https://github.com/user-attachments/assets/c3b85dae-1485-498a-b53f-dc85b3132f5a" />

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -79,27 +97,17 @@ In this example, we will import tiny subset of data from the 1000 Genomes Projec
+
+### Authorization Token
 
 When using the APIs, you need to provide a token for authentication.
```
Replace old screenshot in the following paragraph or provide a link to Getting a Genestack API token article in Advanced Guide

Endalaner replied:
Image was replaced. The article in the advanced guide is also outdated

eeliane replied:
Please, remove information about generating token and provide a link to the existing article, not to create a lot of places with the same information that should be updated separately: quick-start/consumer-api/#generate-a-token

Endalaner replied:
Done

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
-1. **Import a study**
+### Import a study
+
+There are specific endpoints to import specific data types, as listed in the **Swagger API documentation**. 
```
Possibly we could add a link to Swagger here

Endalaner replied:
Done via `[**Swagger API documentation**](/swagger/helper/)`

eeliane replied:
let's provide a link to the certain section with job endpoints

Endalaner replied:
Done

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -127,210 +144,369 @@ with
+```
+
+!!! note "templateId"
+    You can include an optional parameter **"templateId"** to specify which template should be associated with the loaded data. You will need to provide the accession of the desired template, which can be obtained the Template Editor. If the "templateId" parameter is not specified, the default template set for the instance will be used.
````
obtained **from**

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
-  },
-… <three more sample objects here> …
-]
+{
```
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -127,210 +144,369 @@ with
+```
+As soon as the import process will be completed, you will be able to get the sample **groupAccession** by querying the **jobExecId** in **/api/v1/jobs/{jobExecId}/output** endpoint:
+
+```default
````
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -127,210 +144,369 @@ with
-If successful, you will get four run-level expression objects, corresponding to the four columns in the expression matrix:
 
+If successful, you will get the response that contain the **jobExecId** that we will use to get the **groupAccession** using **GET /api/v1/jobs/{jobExecId}/output** endpoint.
 ```default
````
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -127,210 +144,369 @@ with
+```
+As response you will get all the information, including the metadata, for the expression file we have succesfully imported.
+
+```default
````
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -127,210 +144,369 @@ with
+```
+Response will contain the variant data that we imported:
 
 ```default
````
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -368,300 +544,452 @@ This example is similar to the previous one, but demonstrates using library/prep
+  -H 'Genestack-API-Token: <TOKEN>'
+
+```
 ```default
````
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -368,300 +544,452 @@ This example is similar to the previous one, but demonstrates using library/prep
+  -H 'accept: application/json' \
+  -H 'Genestack-API-Token: <TOKEN>'
+```
+```default
````
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -368,300 +544,452 @@ This example is similar to the previous one, but demonstrates using library/prep
+```
+
+This returns similarly to the samples import - [jobExecId](#working-with-the-jobexecid), using the **GET /api/v1/jobs/{jobExecId}/output** endpoint we will get the groupAccession. 
+```default
````
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -368,300 +544,452 @@ This example is similar to the previous one, but demonstrates using library/prep
+    - **measurementSeparator**
+
+If successful, you will get the response that contain the jobExecId that we will use to get the groupAccession using GET /api/v1/jobs/{jobExecId}/output endpoint.
+```default
````
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -368,300 +544,452 @@ This example is similar to the previous one, but demonstrates using library/prep
+  -H 'Genestack-API-Token: <TOKEN>'
+```
 
 ```default
````
"```json"

MikhailAf commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -670,7 +998,10 @@ In the User Interface, you should be able to find your imported study using the
+  -H 'Genestack-API-Token: <TOKEN>'
 ```
 
 Which will return:
````
"```json"

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
 You can also confirm this visually, by going to the **Study Browser** and check that a new study has been created,
 owned by you:
 
 ![image](doc-odm-user-guide/images/empty_study.png)
```
Please, update this image, since its related to the old version of Default/Inception profile

eeliane replied:
@Endalaner  please, add these changes to your PR also

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+---
+
+**PUT /api/v1/jobs/{jobExecId}/restart**
+Restarts a job that has failed or was stopped before completion.
```
@MikhailAf @m-smazhevsky @Endalaner do we need to mention result of this (https://genestack.atlassian.net/browse/ODM-12185) investigation? wdyt guys?

MikhailAf replied:
I agree, that is worth to be mentioned here

eeliane replied:
Sorry, I din't mean that we have to copy info from the ticket and add it to the user guide, it's too technical info and has been written in details for developers, I believe we just need to mention whole idea, and when it works or doesn't work

Endalaner replied:
Added in new commit.

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+
+### Import samples
+
+To import samples, you should use a different endpoint, **/api/v1/jobs/import/samples**:
```
**/api/v1/jobs/import/samples** -> **POST /api/v1/jobs/import/samples**

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+Stops a job that is currently running.
+
+- **Use case**: Use this when you need to cancel a long-running or stuck job.
+- **Endpoint**: `/api/v1/jobs/{jobExecId}/stop`
```
Let's remove this line since we have already mentioned this endpoint two lines above?

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+Restarts a job that has failed or was stopped before completion.
+
+- **Use case**: If a job failed due to a temporary issue, you can restart it using its `jobExecId`.
+- **Endpoint**: `/api/v1/jobs/{jobExecId}/restart`
```
Let's remove this line since we have already mentioned this endpoint two lines above?

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+![sample_added.gif](doc-odm-user-guide/gifs/sample_added.gif)
 
-![image](doc-odm-user-guide/images/sample_added.png)
+Samples from other files can be loaded in the same way. They will be displayed in the Metadata Editor on a separate tab.
```
on a separate tab -> on a separate subtab

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+  -d '{
+  "metadataLink": "https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.gct.tsv",
+  "dataLink": "https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.gct",
+  "numberOfFeatureAttributes": 2,
```
please, remove this line, you need to mention this parameter only for generic data

Endalaner replied:
You are talking about 329?

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -127,210 +144,369 @@ with
+  "dataClass": "Bulk transcriptomics"
+}'
 ```
+The example call in Swagger contain multiple additional fields, that we do not require to be able to import the data. In order to be able to load the data, we will only use *metadataLink*, *dataLink*, *numberOfFeatureAttributes* and *dataClass*.
````
>contain multiple additional fields, that we do not require to be able to import the data

it depends on what format of the file will be used

>In order to be able to load the data, we will only use *metadataLink*, *dataLink*, *numberOfFeatureAttributes* and *dataClass*.

For gct data you don't need numberOfFeatureAttributes parameter. Also metadata parameter can be skipped, it's not required

Endalaner replied:
> The example call in Swagger may include several optional fields that are not required for a successful import. The required fields depend on the format of the file you are importing.
For most tabular formats, the following fields are commonly used:
-metadataLink
-dataLink
-numberOfFeatureAttributes
-dataClass
For GCT-formatted files, the numberOfFeatureAttributes field is not required.
The metadataLink field is optional for all formats — it can be skipped if no associated metadata needs to be linked.

Should we do it like this, or maybe we have a list of requirements for each data type?

eeliane replied:
Here we are talking about expression data, it will be better to provide two examples: for gct file and for generic data file, and explain what parameters should be used and when. Could you please add this information and also mention that only one data class can be used for gct file since as I remember some of our clients tried to use inappropriate data class type.

Endalaner replied:
We are missing a page that lists all the available data classes. Not sure if we should have it here, or somewhere on this page ../supported-formats. wdyt @eeliane @MikhailAf 

>Here we are talking about expression data, it will be better to provide two examples: for gct file and for generic data file, and explain what parameters should be used and when. Could you please add this information and also mention that only one data class can be used for gct file since as I remember some of our clients tried to use inappropriate data class type.

Done

eeliane replied:
As I remember we've provided a link to the list of all available Data Classes, I suggest to mention that only 1 Data Class is available for gct data since at least one of our client tried to import gct with other Data Class, I don't suggest to mention all available Data Classes here :)

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+
+!!! note "Available Parameters"
+    - **metadataLink** - link to a file that contains metadata (.tsv)
+    - **dataLink** - link to a file that contains the data. (.gct)
```
gct is not the only one format for expression data

Endalaner replied:
I have removed the (.gct) to make it easier.

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+    - **dataLink** - link to a file that contains the data. (.gct)
+    - **templateId** - (optional) accession of the template
+    - **previousVersion** - (optional) accession of the previous version of the file. Used to update the existing version of the file.
+    - **numberOfFeatureAttributes** - The number of metadata columns describing each feature (e.g., gene).
```
Let's use description for this parameter from the GUI Uploader

eeliane replied:
The same is actual for all places where you duplicate the same parameters info

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+    - **templateId** - (optional) accession of the template
+    - **previousVersion** - (optional) accession of the previous version of the file. Used to update the existing version of the file.
+    - **numberOfFeatureAttributes** - The number of metadata columns describing each feature (e.g., gene).
+    - **dataClass** - Specify a data class that suits the data set you are importing. You can use [Data Class](import-data-in-odm.md) list as a reference.
```
Let's provide a link to the certain section on the page (#import-data-beta)

eeliane replied:
The same is actual for all places where you duplicate the same parameters info

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+    - **previousVersion** - (optional) accession of the previous version of the file. Used to update the existing version of the file.
+    - **numberOfFeatureAttributes** - The number of metadata columns describing each feature (e.g., gene).
+    - **dataClass** - Specify a data class that suits the data set you are importing. You can use [Data Class](import-data-in-odm.md) list as a reference.
+    - **measurementSeparator**
```
Let's use description for this parameter from the GUI Uploader

eeliane replied:
The same is actual for all places where you duplicate the same parameters info

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
 
-You can then link these run-level expression objects to the corresponding samples. The call below will link
-the expression object GSF282812 to the sample HG00119 (that’s been linked to the study in the previous step):
+You can then link this expression by group (link expression group to sample group).
```
link this expression by group -> link this expression data by group accession

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -127,210 +144,369 @@ with
 ```
 
-1. **Import and link variant data to samples**
+The call below will link the **expression object GSF282812 to the sample HG00119** using the **POST /api/v1/as-curator/integration/link/expression/{sourceId}/to/sample/{targetId}** endpoint:
````
Let's describe that we have two approaches - it's possible to link object to object and group to group, not to confuse our clients, because currently it's not clear why two different endpoints are shown

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+
+### Import and link variant data to samples
+
+Let's repeat the previous step, this time for variant data, ensuring that both expression and variant data are linked to the samples, reinforcing the data model hierarchy where samples are linked to a study, and data types (expression and variant) are linked to samples
```
to samples -> to samples.

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -127,210 +144,369 @@ with
+
+Let's repeat the previous step, this time for variant data, ensuring that both expression and variant data are linked to the samples, reinforcing the data model hierarchy where samples are linked to a study, and data types (expression and variant) are linked to samples
+
+To import the variant data we will use **/api/v1/jobs/import/variant** endpoint:
```
**/api/v1/jobs/import/variant** -> **POST /api/v1/jobs/import/variant**

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -368,300 +544,452 @@ This example is similar to the previous one, but demonstrates using library/prep
+6. **Import and link expression data to the library file (note linking to library file instead of samples)**
+7. **Check that you can query the relationships between objects**
+
+### Authorization Token
```
Let's skip here repeating of the same information mentioned above and just provide a link to the needed section (I mean that we repeat again this information: Generate an API token, Use Access Token, Import a study, Import samples, Link samples to study, Import gene expression data (with mentioning again all parameters)

Endalaner replied:
I had an idea of changing the structure of this article - by providing the basic steps (token, study and samples) in the first unified part and then separating into 2 block with different approach. The only issue is that originally we are using 2 different test data sets, and we probably will need to save this information.

I have cut the description and left a note pointing to the description in the beginning. WDYT?

eeliane replied:
I think that we don't have to repeat the same information again and again, it makes this articles pretty big and difficult for reading. I believe that we can mention importing libraries/preparations and linking to samples without creating a separate dataset with new study and samples, we can update the dataset that has been used at the beginning. Additionally we can skip mentioning several times approach with importing expression data, its required and optional parameters and just describe whole flow in details one time, and just add information that the same works when we'd like to import expression data to libs/preps, just mention required columns in gct file, what attributes can be used for linking, and what endpoints. And also it'd be good to mention one time how gct and generic data can be imported, but it seems to me that I've already mentioned this somewhere in comments. in this way you'll be able simplify this article without adding the same massive instructions several times. WDYT?

Endalaner replied:
Implemented

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -368,300 +544,452 @@ This example is similar to the previous one, but demonstrates using library/prep
 }
 ```
 
 You can then link this expression group object to the library object. :
````
Please, let's rephrase it to "link expression group to library group"

Commit:
- Updated the section headers and content to improve readability and organization.
- Enhanced descriptions of data types and their relationships during the import process.
- Clarified the limitations regarding attached files and provided workarounds.
- Improved the explanation of the "templateId" parameter and its usage.
- Added detailed behavior descriptions for different file types during job execution.
- Updated various examples and code snippets for consistency and accuracy.

eeliane committed:
Merge branch 'develop' into Import-data-API-update

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -368,300 +593,420 @@ This example is similar to the previous one, but demonstrates using library/prep
+### Import samples
 
-To import samples, you should use a different endpoint, *sampleCurator*:
+To import samples, you should use a different job endpoint, **/api/v1/jobs/import/samples**:
```
/api/v1/jobs/import/samples -> POST /api/v1/jobs/import/samples

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -368,300 +593,420 @@ This example is similar to the previous one, but demonstrates using library/prep
+
+!!! note "Available Parameters"
+    - **metadataLink** - link to a file that contains metadata (.tsv)
+    - **dataLink** - link to a file that contains the data. (.gct)
```
gct is not the only one format for expression data

Commit:
Article structure change, added info on attached files, facs, preparations etc

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
+- Expression can be linked to library or preparations
+- Attached files are linked directly to a study
+
+Importing data has two stages. First, you import studies, samples, and data separately. Then, you link them together: a study can be linked to multiple samples, and a sample can be linked to multi-omics or other types of data. The **Sample Source ID** is used as the default linking key. You can choose another attribute from the template for linking data to samples. The data model and how it looks ...
```
Please, rephrase this to mention proper order

>a study can be linked to multiple samples, and a sample can be linked to multi-omics or other types of data


we link samples to study and multi-omics data can be linked to samples, not vice versa

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
+
+Importing data has two stages. First, you import studies, samples, and data separately. Then, you link them together: a study can be linked to multiple samples, and a sample can be linked to multi-omics or other types of data. The **Sample Source ID** is used as the default linking key. You can choose another attribute from the template for linking data to samples. The data model and how it looks ...
+
+In addition to core data types, **libraries** and **preparations** require special handling. Unlike omics data, they cannot be linked using arbitrary attributes — they must include their own unique ID that corresponds to the **Sample Source ID**.
```
>they must include their own unique ID that corresponds to the Sample Source ID.

Please, mention proper required columns in the file and what column with data will be used for linking to samples. You have this information below, but here is written something strange

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
````diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
+  "createTime": "2025-04-16 13:47:17"
+}
 ```
+As soon as the import process will be completed, you will be able to get the sample **groupAccession** by querying the **jobExecId** in `GET /api/v1/jobs/{jobExecId}/output` endpoint:
````
>GET /api/v1/jobs/{jobExecId}/output

Could you please recheck what formatting we used for endpoints? Looks like currently you have two approach, let's use just one in whole article

Endalaner replied:
Done

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
+#### Libraries to Samples
 
-To import the variant data:
+You can link the **library group file** to the **samples group** using the endpoint `POST /api/v1/as-curator/integration/link/library/group/{sourceId}/to/sample/group/{targetId}`, along with the **accession** returned when importing the samples.
```
library group file -> library group

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
+
+#### Preparations to Samples
+
+You can link the **preparation group file** to the **samples group** using the endpoint `POST /api/v1/as-curator/integration/link/preparation/group/{sourceId}/to/sample/group/{targetId}`, along with the **accession** returned when importing the samples.
```
preparation group file -> preparation group

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
+Retrieves the current status and metadata of a specific job execution.
+
+- **Use case**: Use this to monitor the progress of an import job using its `jobExecId`.
+- **Endpoint**: `/api/v1/jobs/{jobExecId}/info`
```
/api/v1/jobs/{jobExecId}/info -> GET /api/v1/jobs/{jobExecId}/info (minor)

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
+Retrieves the output of a completed job, including the accession of the generated study.
+
+- **Use case**: Use this after a job has completed to get the final result and study accession.
+- **Endpoint**: `/api/v1/jobs/{jobExecId}/output`
```
/api/v1/jobs/{jobExecId}/output ->  GET /api/v1/jobs/{jobExecId}/output (minor)

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
 
-To import samples, you should use a different endpoint, *sampleCurator*:
+!!! note "Linking via Sample Source ID"
+    When linking signal data to preparations or libraries, the linking must be made through the default attribute `Sample Source ID`.
```
>When linking signal data to preparations or libraries, the linking must be made through the default attribute `Sample Source ID`


Please, fix this to the actual and correct behavior (critical)

Endalaner replied:
Done

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
 
-You can also confirm this visually, by going to the **Study Browser** and check that a new study has been created,
-owned by you:
+In this example, we link an expression group to a sample group using `POST /api/v1/as-curator/integration/link/{sourceType}/group/{sourceId}/to/{targetType}/group/{targetId}`. 
```
>In this example, we link an expression group to a sample group using

Based on example of gct file provided above, please, use as a example of linkage endpoint to link expression data to libraries, not to samples or considering that you mentioned below linkage to samples, there's an option to update/change your example of gct file

Endalaner replied:
I have added TSV that can be linked to a library, provided an example call to link expression to a library group.
We now have 3 types of expression data (GCT, TSV and TSV with lib link attr).

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
-| LIB2         | SRR6441196            | NGS                    | RNA-Seq-1         |
+## Signal Data Import
+
+### Expression data
```
Let's please also add one example of expression data in tsv format, so we will be able to cover clients' questions about importing so called generic data, and this example will show using of parameters 'numberOfFeatureAttributes', 'measurementSeparator'. It's even not necessary to link this file to any group

eeliane replied:
https://github.com/genestack/user-docs/pull/120#discussion_r2086824659

Endalaner replied:
Done. Added file example, description, example call and link to the page that describes the tabular data type.

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
+| HG00119  | Total events/Lymphocytes/Single Cells        | Counts      |              | 177879  |
+| HG00119  | Total events/Lymphocytes/Single Cells        | Percentage  |              | 97.4    |
+
+- [Test_FACS_Signals.facs.csv](https://bio-test-data.s3.us-east-1.amazonaws.com/odm/user-guide/Test_FACS_Signals.facs.csv), a comma-separated file that describes metadata for FACS data, including source and field definitions.
```
>comma-separated


We don't support a comma-separated files, please, fix it

Could you please compare it with the description for expression and variant metadata file and use more or less identical description?

Endalaner replied:
My bad, fixed.

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
+
+To import and link attached file to a study we will use `POST /api/v1/jobs/import/file` endpoint.
+
+The example call contain link to a file, accession of a study that the file will be linked to and a file type for the imported file.
```
file type for the imported file -> Data Class for the imported file.

Let's please mention that user can use any available Data Class for attached file

+ Let's please mention that here study accession is a mandatory parameter and in general is needed to attach file to the certain study without additional linking

Endalaner replied:
>!!! note "Mandatory fields" 
> Please note that `dataLink`, `studyAccession` and `dataClass` are mandatory fields and cannot be skipped.

Added a note with a list of mandatory fields.

> and in general is needed to attach file to the certain study without additional linking

The user will receive an error message if he will try to import the data with any of these fields missing. I propose to skip this part, since it might confuse them and they might try to import the file without the study accession to try and link it later.

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
-Once you’ve created and linked the study, sample, library and expression objects you can do integration-aware queries via both the User Interface and APIs.
+Once you've created and linked the study, sample, library and expression objects you can do integration-aware queries via both the User Interface and APIs.
 
 In the User Interface, you should be able to find your imported study using the study, sample, and signal filters.
```
What's about libraries and preparations filters?

Endalaner replied:
Added.

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
 
 In the User Interface, you should be able to find your imported study using the study, sample, and signal filters.
 
 To do this via APIs, you can use the integration/omics endpoint to filter across studies, samples, libraries and signals,and retrieve a specific object type. For example, to get metadata about the samples associated with library LIB1:
```
libraries and signals,and -> libraries and signals, and

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,740 @@ you can try. Please note that you need to be a member of the curator group in OD
-When using the APIs, you need to provide an API token for authentication.
-You can generate this by going to your **profile**, which can be found by clicking your username at the top right corner
-of the User Interface, or from the Dashboard.
+!!! note "Available Parameters"
```
Maybe there's a sense to mention available and mandatory parameters for other job endpoints, especially for attached files

Endalaner replied:
Good point. Added for the Attached Endpoint.
I'll check what restrictions we have for others

Commit:
Enhanced section on linking entities, clarified examples for expression data and FACS files, and added details on mandatory fields for data import.

Endalaner committed:
Merge branch 'develop' into Import-data-API-update

Commit:
Clarifying linking processes for studies, samples, libraries, and preparations. Update examples for API endpoints and emphasize mandatory fields for data import, including data class specifications for GCT and TSV files.

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,812 @@ you can try. Please note that you need to be a member of the curator group in OD
+- **Preparations metadata**: metadata describing how samples were prepared prior to data generation, applicable to proteomics, transcriptomics, and other data types.
+- **Attached Files**: Supplement your study by attaching related research materials like PDF, XLSX, DOCX, PPTX files, images, and more. Please note, contents of these attached files won't be indexed or made searchable.
 
 Once imported, studies, samples, and data metadata will be queryable and editable from both the User Interface and APIs, whilst the signal data will only queryable via APIs.
```
will only queryable via APIs -> will only be queryable via APIs?

Commit:
Added a new tabular dataset example in TSV format linked to libraries, clarified API endpoints for linking expression groups to libraries and preparations, and updated examples for linking processes in the user guide.

Commit:
minor fixes

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,832 @@ you can try. Please note that you need to be a member of the curator group in OD
-1. **Import a study**
+!!! note "Available Parameters"
+    - **dataLink** - link to a file that contains the data.
+    - **dataClass** - Specify a data class that suits the data set you are importing. You can use [Data Class](import-data-in-odm.md#import-data-beta) list as a reference.
```
import-data-in-odm/#import-data-beta -> import-data-in-odm/#data-type-data-class

eeliane commented on docs/user-guide/doc-odm-user-guide/import-data-using-api.md:
```diff
@@ -7,675 +17,832 @@ you can try. Please note that you need to be a member of the curator group in OD
+| HG00119  | Total events/Lymphocytes/Single Cells        | Percentage  |              | 97.4    |
 
-1. **Import and link gene expression data to libraries**
+- [Test_FACS_Signals.facs.csv](https://bio-test-data.s3.us-east-1.amazonaws.com/odm/user-guide/Test_FACS_Signals.facs.csv),a tab-separated file that describes the FACS data.
```
,a tab-separated file that describes the FACS data -> , a tab-separated file that describes the FACS data

Commit:
minor edits
